### DISCUSSION TITLE
BBCode syntax highlighting

### DISCUSSION THREAD

**Discussion by seshoumara** [godotengine/godot-proposals#9258] [Editor]
When working with RitchTextLabel and BBCode, especially when using multiple tags, it's incredibly difficult to tell where a tag starts and where it ends. It's also hard to distinguish between tags, options, values and regular text. As such, I think having a syntax highlighting for BBCode would greatly improve its ease of use. See current readability in screenshot below:

![image](https://github.com/godotengine/godot-proposals/assets/46658810/1d44c7a8-0630-4404-b19b-64fad1585784)

One last suggestion is to highlight custom BBCode tags differently from the default tags, if possible.

The implementation is up to you, from reusing existing highlight colors of regular code from the Text Editor setting or adding new colors with BBCode prefix/suffix in their name or something else. I don't have experience to tell if this should be in the core or done using GDExtension.

**Comment by Mickeon** [comment]
In theory it could be done with an addon, by the way. See also https://docs.godotengine.org/en/stable/classes/class_editorproperty.html and https://docs.godotengine.org/en/stable/classes/class_editorinspector.html

**Comment by MrEliptik** [comment]
I agree, editing BBcode is a pain a could be made better.

I would also suggest autocomplete. If you write a open tag, it could close automatically, just like () in gdscript.
I also think the possibility to highlight text and apply a tag with a shortcut or right click could be awesome. I don't know if this should be another proposal?

**Reply by Mickeon** [reply]
If you want to go more in detail about it feel free to make another proposal. I feel like what you're suggesting are straight upgrades with not much to discuss.

A few questions remain, though.
How does the Inspector deduce that the text being written is using BBCode? Probably with a PROPERTY_HINT.
What colours to use for the syntax highlighting? Having additional custom colors in the Editor Settings for it may be a bit overkill.

**Reply by seshoumara** [reply]
You know BBCode is being written because the 'Enabled' checkbox is on. See image at the top. As for the colors, I don't think it is overkill. It would be clear what they're for and you don't have the coupling where one color is responsible for two highlights, regular code and BBCode.

![image](https://github.com/godotengine/godot-proposals/assets/46658810/00155cae-1db0-47b6-acd8-495c6bc2b3f0)

In the Text Editor settings, on Highlighting (see above), there is a sort of section for GDScript-specific things. I can envision another section called 'BBCode' with all the respective colors underneath.

**Comment by MarioLiebisch** [comment]
Even if it could be done in a plugin, I'd say this is a big enough QoL change to warrant it being in the base engine (especially considering it doesn't need a new component either).

**Reply by LauraWebdev** [reply]
Especially considering this is a change that benefits everyone. I don't think making devs find and install a plugin for each of their project for this is the right way.

**Comment by Calinou** [comment]
We have `PROPERTY_HINT_EXPRESSION` that is used for `AnimationNodeStateMachineTransition.advance_expression`. This property hint makes text show up with a fixed-width code font instead of the main font and uses standard syntax highlighting (based on braces, keywords – but not GDScript-specific):

![image](https://github.com/godotengine/godot-proposals/assets/180032/4907e431-09c3-4314-b8fe-7af19a9bd7a0)

We can probably add `PROPERTY_HINT_BBCODE` that does something similar (perhaps without the code font though), implement a BBCode syntax highlighter and toggle it based on the value of `bbcode_enabled` in the RichTextLabel node.

The implementation of `PROPERTY_HINT_EXPRESSION` is here: https://github.com/godotengine/godot/blob/da945ce6266ce27ba63b6b08dc0eb2414594f7cb/editor/editor_properties.cpp#L3648-L3650

https://github.com/godotengine/godot/blob/da945ce6266ce27ba63b6b08dc0eb2414594f7cb/editor/editor_properties.cpp#L242-L247

We'd need to change EditorPropertyMultilineText's constructor so it accepts an enum with a language type instead of a boolean, but this should be feasible as this API isn't exposed to scripting.

Developing the highlighter itself might prove nontrivial though. You can see my attempt at creating a [XML syntax highlighter](https://github.com/Calinou/godot/tree/editor-add-xml-highlighting) which didn't quite look perfect, although it was still better than nothing. I'd like different tags to be highlighted differently based on their depth or purpose (color, size, alignment), rather than having all tags be highlighted with the same color.

**Comment by shoyguer** [comment]
<img width="1577" height="575" alt="image" src="https://github.com/user-attachments/assets/887e12ca-b931-4f0a-b9de-d6b8528f226f" />

I'm testing it, and honestly, it makes things way more readable.
I'll probably open a PR for a contribution in a few days about this proposal.

**Reply by MarioLiebisch** [reply]
Looks great although I'd map the bracket colors to the names, too. In your case both should be red.

**Reply by seshoumara** [reply]
Much better readability, looks great! Does a custom bbcode tag (unknown tag) have its own color? Are these colors configurable from somewhere? I'd use it as it is, the mentioned things are nice to have.

**Reply by shoyguer** [reply]
@MarioLiebisch just fixed it! Thanks for the idea 😄

**Reply by shoyguer** [reply]
@seshoumara 
The colors are configurable. I added a dedicated BBCode section under Editor Settings / Text Editor / Theme / Highlighting / BBCode, check it out:
<img width="669" height="399" alt="image" src="https://github.com/user-attachments/assets/76ffd2ba-a26b-423f-a2ae-8678d06400fa" />

I also already added syntax highlight on Godot Strings on normal code editor.
<img width="1399" height="120" alt="image" src="https://github.com/user-attachments/assets/74621452-0304-48e0-ac4b-4a4923b99f8a" />

For custom/unknown tags, they currently use the same color as built-in tags. I'm currently thinking of a way to solve this, since maybe people just want to add some text inside brackets, and Godot will automatically highlight that, and this might be inconvenient.

**Reply by shoyguer** [reply]
<img width="489" height="319" alt="image" src="https://github.com/user-attachments/assets/86aaf6d7-5c9c-439b-964b-fc8035e58e53" />

Added autocompletion

**Comment by dalexeev** [comment]
I have a similar [addon](https://github.com/dalexeev/godot-inspector-code-highlighter) for GDScript syntax highlighting. I also have an unpublished addon for BBCode editing:

<img width="250" height="174" alt="" src="https://github.com/user-attachments/assets/2b2ce2a5-0c76-4a7b-b45c-3201e126776b" />

<img width="996" height="888" alt="" src="https://github.com/user-attachments/assets/25d117db-5ab6-478f-8d4a-dfd2c1548475" />

I made this prototype because a friend asked. But my friend preferred to use a separate utility he created himself. I wanted to add autocompletion for tags and parameters, but I didn't finish the addon. If you are interested, I can publish this addon and possibly improve it later.

**Reply by shoyguer** [reply]
Just checked the Addon, very interesting stuff you have there!
Also very useful indeed honestly

**Comment by elliotfontaine** [comment]
Linking @shoyguer's PR:
https://github.com/godotengine/godot/pull/117803
